File: lab4_6_1/1.py
import heapq
import sys

class Node:
    def __init__(self, node_id, h_cost):
        self.id = node_id
        self.h_cost = h_cost

    def __lt__(self, other):
        return self.h_cost < other.h_cost

def best_first_search(n, adj, start, target, h_values):
    pq = [Node(start, h_values[start])]
    visited = [False] * (n + 1)
    parent = [-1] * (n + 1)
    visited[start] = True

    found = False

    while pq:
        current = heapq.heappop(pq)
        u = current.id

        print(f"Visiting Node: {u} | Heuristic: {current.h_cost}")

        if u == target:
            print(f"\nTarget Found at Node {u}")
            found = True
            break

        if u in adj:
            for v in adj[u]:
                if not visited[v]:
                    visited[v] = True
                    parent[v] = u
                    heapq.heappush(pq, Node(v, h_values[v]))

    if found:
        print("Path to Target: ", end="")
        path = []
        crawl = target
        while crawl != -1:
            path.append(crawl)
            crawl = parent[crawl]
        path.reverse()
        print(" -> ".join(map(str, path)))
    else:
        print("Target is unreachable")

def main():
    try:
        input_data = sys.stdin.read().split()
        if not input_data: return
        it = iter(input_data)
        
        n = int(next(it))
        m = int(next(it))
        
        adj = {}
        for _ in range(m):
            u = int(next(it))
            v = int(next(it))
            if u not in adj: adj[u] = []
            if v not in adj: adj[v] = []
            adj[u].append(v)
            adj[v].append(u)

        start = int(next(it))
        target = int(next(it))

        h_values = []
        for _ in range(n + 1):
            h_values.append(int(next(it)))

        best_first_search(n, adj, start, target, h_values)
    except (StopIteration, ValueError, EOFError):
        pass

if __name__ == "__main__":
    main()

File: lab4_6_1/1.cpp
#include<bits/stdc++.h>

using namespace std;

struct Node{
    int id;
    int h_cost;

    Node(int i, int h) : id(i), h_cost(h) {}

    bool operator >(const Node& other) const {
        return h_cost > other.h_cost;
    }
};

void bestFirstSearch(int n, vector<vector<int>>& adj, int start, int target, vector<int>& h_values) {
    priority_queue<Node, vector<Node>, greater<Node>> pq;

    vector<bool> visited(n + 1, false);
    vector<int> parent(n + 1, -1);

    pq.push(Node(start, h_values[start]));
    visited[start] = true;

    bool found = false;

    while (!pq.empty()) {
        Node current = pq.top();
        pq.pop();
        int u = current.id;

        cout << "Visiting Node: " << u << " | Heuristic: " << current.h_cost << endl;

        if (u == target) {
            cout << "\nTarget Found at Node " << u << endl;
            found = true;
            break;
        }

        for (int v : adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                parent[v] = u;
                pq.push(Node(v, h_values[v]));
            }
        }
    }

    if (found) {
        cout << "Path to Target: ";
        stack<int> path;
        int crawl = target;

        while (crawl != -1) {
            path.push(crawl);
            crawl = parent[crawl];
        }

        while (!path.empty()) {
            int p = path.top();
            path.pop();
            cout << p;
            if (!path.empty()) cout << " -> ";
        }
        cout << "\n";
    } else {
        cout << "Target is unreachable\n";
    }
}

int main() {
    int n, m;
    cout << "Enter no. of node & edges : ";
    cin >> n >> m;

    vector<vector<int>> adj(n + 1);

    cout << "enter src & dest of each edge :\n";
    int u, v;
    for(int i = 0; i < m; i++) {
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    int start, target;
    cout << "Enter Start Node and Target Node: ";
    cin >> start >> target;

    vector<int> h_values(n + 1);
    cout << "Enter heuristic value for each node (from 0 to " << n << "):\n";
    for(int i = 0; i <= n; i++) {
        cin >> h_values[i];
    }

    bestFirstSearch(n, adj, start, target, h_values);

    return 0;
}